File: frontend/src/api/desensitize.ts
import request from '@/utils/request'

export interface Entity {
  type: string
  text: string
  start: number
  end: number
  score: number
  source: string
  replacement?: string
  replacement_method?: string
  context_label?: string
  context_role?: string
  canonical_key?: string
  canonical_role?: string
  group_id?: string
  group_label?: string
  needs_review?: boolean
  review_reason?: string
}

export interface AnalyzeResponse {
  task_id: string
  filename: string
  text: string
  entities: Entity[]
  statistics: Record<string, any>
  metadata: Record<string, any>
  llm_model?: string
  llm_strategy?: string
  llm_strategy_label?: string
  anonymization_strategy?: string
  anonymization_strategy_label?: string
}

export interface TaskStatus {
  task_id: string
  filename?: string
  status: string
  progress: number
  message?: string
  error_message?: string
  created_at: string
}

export interface BatchFileItem {
  item_id: string
  filename: string
  relative_path: string
  status: string
  progress: number
  message?: string
  error_message?: string
  entities_count: number
  output_filename?: string
  mapping_output_filename?: string
  output_file_type?: string
  mapping_output_file_type?: string
  preserves_format: boolean
  warning?: string
  download_url?: string
  mapping_download_url?: string
  metadata: Record<string, any>
}

export interface BatchTaskStatus {
  batch_id: string
  folder_name: string
  output_folder_name?: string
  status: string
  progress: number
  message?: string
  error_message?: string
  file_count: number
  completed_count: number
  succeeded_count: number
  failed_count: number
  created_at: string
  items: BatchFileItem[]
}

export interface BatchResult {
  batch_id: string
  folder_name: string
  output_folder_name?: string
  status: string
  progress: number
  message?: string
  error_message?: string
  file_count: number
  completed_count: number
  succeeded_count: number
  failed_count: number
  archive_download_url?: string
  archive_filename?: string
  items: BatchFileItem[]
}

export interface DesensitizeRequest {
  task_id: string
  entities: Entity[]
  config?: Record<string, any>
  llm_model?: string
  anonymization_strategy?: string
  desensitize_mode?: string
  page_session_id?: string
  async_mode?: boolean
}

export interface DesensitizeResponse {
  task_id: string
  status: string
  anonymized_text?: string
  entities: Entity[]
  metadata?: Record<string, any>
  download_url?: string
  mapping_download_url?: string
  output_filename?: string
  mapping_output_filename?: string
  output_file_type?: string
  mapping_output_file_type?: string
  preserves_format?: boolean
  llm_assisted?: boolean
  llm_model?: string
  llm_strategy?: string
  llm_strategy_label?: string
  anonymization_strategy?: string
  anonymization_strategy_label?: string
  warning?: string
  message?: string
}

export interface EngineInfo {
  version: string
  architecture: string
  llm_backend: string
  desensitize_mode?: string
  llm_model: string
  supported_entities: string[]
  supported_operators: string[]
  statistics: Record<string, any>
}

export interface LLMModelOption {
  name: string
  installed: boolean
  is_default: boolean
  strategy_key: string
  strategy_label: string
  strategy_description: string
  tier: string
  supports_precision_review: boolean
  supports_vision: boolean
  recommended_for: string[]
  role?: string
  memory_tier?: string
  local_path?: string
}

export interface LLMModelListResponse {
  backend: string
  default_model?: string
  service_available: boolean
  models: LLMModelOption[]
  profile?: string
  primary_models?: string[]
  review_models?: string[]
}

export interface RuntimeStatusResponse {
  backend: string
  platform: string
  ready: boolean
  ollama_install_detected: boolean
  ollama_path?: string
  service_available: boolean
  required_model?: string
  required_model_installed: boolean
  available_processing_models: string[]
  preferred_processing_model?: string
  default_model?: string
  installer_hint: string
  download_hint: string
  recommended_action: string
  desensitize_mode?: string
  primary_models_ready?: boolean
  review_model_installed?: boolean
  review_model_loaded?: boolean
  estimated_memory_tier?: string
  analysis_worker_process?: boolean
  analysis_stage_isolation?: boolean
  analysis_worker_timeout?: number
}

export const uploadAndAnalyze = (
  file: File,
  useLlm = true,
  useCustom = true,
  llmModel?: string,
  anonymizationStrategy?: string,
  desensitizeMode?: string,
  pageSessionId?: string,
  wpsDocxTemplate?: File | null
) => {
  const formData = new FormData()
  formData.append('file', file)
  formData.append('use_llm', String(useLlm))
  formData.append('use_custom', String(useCustom))
  formData.append('async_mode', 'true')
  if (llmModel) {
    formData.append('llm_model', llmModel)
  }
  if (anonymizationStrategy) {
    formData.append('anonymization_strategy', anonymizationStrategy)
  }
  if (desensitizeMode) {
    formData.append('desensitize_mode', desensitizeMode)
  }
  if (pageSessionId) {
    formData.append('page_session_id', pageSessionId)
  }
  if (wpsDocxTemplate) {
    formData.append('wps_docx_template', wpsDocxTemplate)
  }

  return request.post<any, TaskStatus>('/desensitize/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

export const uploadAndProcessFolder = (
  files: File[],
  relativePaths: string[],
  folderName: string,
  useLlm = true,
  useCustom = true,
  llmModel?: string,
  anonymizationStrategy?: string,
  operatorConfig?: Record<string, any>,
  desensitizeMode?: string,
  pageSessionId?: string
) => {
  const formData = new FormData()
  files.forEach((file, index) => {
    formData.append('files', file, file.name)
    formData.append('relative_paths', relativePaths[index] || file.name)
  })
  formData.append('folder_name', folderName)
  formData.append('use_llm', String(useLlm))
  formData.append('use_custom', String(useCustom))
  if (llmModel) {
    formData.append('llm_model', llmModel)
  }
  if (anonymizationStrategy) {
    formData.append('anonymization_strategy', anonymizationStrategy)
  }
  if (operatorConfig && Object.keys(operatorConfig).length > 0) {
    formData.append('operator_config_json', JSON.stringify(operatorConfig))
  }
  if (desensitizeMode) {
    formData.append('desensitize_mode', desensitizeMode)
  }
  if (pageSessionId) {
    formData.append('page_session_id', pageSessionId)
  }

  return request.post<any, BatchTaskStatus>('/desensitize/batch/upload', formData, {
    headers: { 'Content-Type': 'multipart/form-data' }
  })
}

export const processDesensitize = (data: DesensitizeRequest) => {
  return request.post<any, DesensitizeResponse>('/desensitize/process', data)
}

export const processDesensitizeAsync = (data: DesensitizeRequest) => {
  return request.post<any, TaskStatus>('/desensitize/process', {
    ...data,
    async_mode: true
  })
}

export const getAnalyzeResult = (taskId: string) => {
  return request.get<any, AnalyzeResponse>(`/desensitize/result/${taskId}`)
}

export const getProcessedResult = (taskId: string) => {
  return request.get<any, DesensitizeResponse>(`/desensitize/processed-result/${taskId}`)
}

export const downloadResult = (taskId: string) => `/api/v1/desensitize/download/${taskId}`

export const downloadMappingResult = (taskId: string) =>
  `/api/v1/desensitize/download/mapping/${taskId}`

export const getTaskStatus = (taskId: string) => {
  return request.get<any, TaskStatus>(`/desensitize/status/${taskId}`)
}

export const getBatchTaskStatus = (batchId: string) => {
  return request.get<any, BatchTaskStatus>(`/desensitize/batch/status/${batchId}`)
}

export const getBatchResult = (batchId: string) => {
  return request.get<any, BatchResult>(`/desensitize/batch/result/${batchId}`)
}

export const downloadBatchArchive = (batchId: string) =>
  `/api/v1/desensitize/batch/download/${batchId}`

export const downloadBatchItem = (batchId: string, itemId: string) =>
  `/api/v1/desensitize/batch/download/${batchId}/${itemId}`

export const downloadBatchItemMapping = (batchId: string, itemId: string) =>
  `/api/v1/desensitize/batch/download/${batchId}/${itemId}/mapping`

export const getEngineInfo = (desensitizeMode?: string) => {
  return request.get<any, EngineInfo>('/desensitize/info', {
    params: { desensitize_mode: desensitizeMode }
  })
}

export const getAvailableModels = (desensitizeMode?: string) => {
  return request.get<any, LLMModelListResponse>('/desensitize/models', {
    params: { desensitize_mode: desensitizeMode }
  })
}

export const getRuntimeStatus = (desensitizeMode?: string) => {
  return request.get<any, RuntimeStatusResponse>('/desensitize/runtime-status', {
    params: { desensitize_mode: desensitizeMode }
  })
}

File: backend/app/schemas/desensitize.py
"""Pydantic schemas for the desensitization APIs."""

from datetime import datetime
from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class Entity(BaseModel):
    type: str = Field(..., description="Entity type")
    text: str = Field(..., description="Matched text")
    start: int = Field(..., description="Start offset")
    end: int = Field(..., description="End offset")
    score: float = Field(default=0.9, description="Confidence score")
    source: str = Field(default="unknown", description="Recognizer source")
    replacement: Optional[str] = Field(default=None, description="Prepared replacement text")
    replacement_method: Optional[str] = Field(default=None, description="Replacement generation method")
    context_label: Optional[str] = Field(default=None, description="Nearby field label")
    context_role: Optional[str] = Field(default=None, description="Inferred contract role")
    canonical_key: Optional[str] = Field(default=None, description="LLM-resolved canonical identity key")
    canonical_role: Optional[str] = Field(default=None, description="LLM-resolved canonical role")
    group_id: Optional[str] = Field(default=None, description="Derived canonical group ID")
    group_label: Optional[str] = Field(default=None, description="Human-readable group label")
    needs_review: bool = Field(default=False, description="Whether this entity should be reviewed manually")
    review_reason: Optional[str] = Field(default=None, description="Reason the entity should be reviewed")
    metadata: Dict = Field(default_factory=dict, description="Recognizer metadata and identity hints")


class AnalyzeResponse(BaseModel):
    task_id: str = Field(..., description="Task ID")
    filename: str = Field(..., description="Original filename")
    text: str = Field(..., description="Extracted source text")
    entities: List[Entity] = Field(default_factory=list, description="Detected entities")
    statistics: Dict = Field(default_factory=dict, description="Entity statistics")
    metadata: Dict = Field(default_factory=dict, description="Document metadata")
    llm_model: Optional[str] = Field(default=None, description="LLM model used during analysis")
    llm_strategy: Optional[str] = Field(default=None, description="LLM strategy key used during analysis")
    llm_strategy_label: Optional[str] = Field(default=None, description="LLM strategy label used during analysis")
    anonymization_strategy: Optional[str] = Field(default=None, description="Replacement strategy key")
    anonymization_strategy_label: Optional[str] = Field(default=None, description="Replacement strategy label")


class DesensitizeRequest(BaseModel):
    task_id: str = Field(..., description="Task ID")
    entities: List[Entity] = Field(..., description="Entities confirmed by the user")
    config: Optional[Dict] = Field(default=None, description="Operator configuration")
    llm_model: Optional[str] = Field(default=None, description="Requested LLM model")
    anonymization_strategy: Optional[str] = Field(default=None, description="Requested replacement strategy")
    desensitize_mode: Optional[str] = Field(default=None, description="Requested processing workflow")
    page_session_id: Optional[str] = Field(default=None, description="Frontend page session ID")
    async_mode: bool = Field(default=False, description="Whether to run anonymization/export in the background")


class DesensitizeResponse(BaseModel):
    task_id: str = Field(..., description="Task ID")
    status: str = Field(..., description="Task status")
    anonymized_text: Optional[str] = Field(default=None, description="Preview text")
    entities: List[Entity] = Field(default_factory=list, description="Entities with prepared replacements")
    metadata: Dict = Field(default_factory=dict, description="Processing metadata")
    download_url: Optional[str] = Field(default=None, description="Download URL")
    mapping_download_url: Optional[str] = Field(default=None, description="Mapping document download URL")
    output_filename: Optional[str] = Field(default=None, description="Generated file name")
    mapping_output_filename: Optional[str] = Field(default=None, description="Generated mapping file name")
    output_file_type: Optional[str] = Field(default=None, description="Generated file type")
    mapping_output_file_type: Optional[str] = Field(default=None, description="Generated mapping file type")
    preserves_format: bool = Field(default=False, description="Whether the original format was preserved")
    llm_assisted: bool = Field(default=False, description="Whether the LLM participated in processing")
    llm_model: Optional[str] = Field(default=None, description="LLM model used during processing")
    llm_strategy: Optional[str] = Field(default=None, description="LLM strategy key used during processing")
    llm_strategy_label: Optional[str] = Field(default=None, description="LLM strategy label used during processing")
    anonymization_strategy: Optional[str] = Field(default=None, description="Replacement strategy key")
    anonymization_strategy_label: Optional[str] = Field(default=None, description="Replacement strategy label")
    warning: Optional[str] = Field(default=None, description="Export warning")
    message: Optional[str] = Field(default=None, description="Status message")


class TaskStatus(BaseModel):
    task_id: str = Field(..., description="Task ID")
    filename: Optional[str] = Field(default=None, description="Original filename")
    status: str = Field(..., description="Task status")
    progress: int = Field(default=0, description="Progress from 0 to 100")
    message: Optional[str] = Field(default=None, description="Status message")
    error_message: Optional[str] = Field(default=None, description="Terminal error message")
    created_at: datetime = Field(default_factory=datetime.now, description="Creation time")


class PdfWordAuditStatus(BaseModel):
    audit_id: str = Field(..., description="Audit task ID")
    filename: Optional[str] = Field(default=None, description="Original PDF filename")
    template_filename: Optional[str] = Field(default=None, description="WPS DOCX template filename")
    status: str = Field(..., description="Task status")
    progress: int = Field(default=0, description="Progress from 0 to 100")
    message: Optional[str] = Field(default=None, description="Status message")
    error_message: Optional[str] = Field(default=None, description="Terminal error message")
    created_at: datetime = Field(default_factory=datetime.now, description="Creation time")


class PdfWordAuditCorrection(BaseModel):
    id: str = Field(..., description="Correction candidate ID")
    wps_unit_id: str = Field(..., description="Matched WPS text unit ID")
    page_no: Optional[int] = Field(default=None, description="Estimated page number")
    old_text: str = Field(default="", description="WPS text")
    new_text: str = Field(default="", description="OCR suggested text")
    action: str = Field(default="review", description="Correction action")
    confidence: float = Field(default=0.0, description="OCR confidence")
    alignment_score: float = Field(default=0.0, description="Alignment score")
    reason: str = Field(default="", description="Reason for action")
    comment_text: str = Field(default="", description="DOCX comment body")
    sensitive_low_priority: bool = Field(default=False, description="Whether this likely belongs to a desensitization field")


class PdfWordAuditFinding(BaseModel):
    id: str = Field(..., description="Finding ID")
    severity: str = Field(default="medium", description="Finding severity")
    category: str = Field(default="substitution", description="Difference category")
    page_no: Optional[int] = Field(default=None, description="Page number")
    wps_text: str = Field(default="", description="WPS converted text")
    suggested_text: str = Field(default="", description="Preferred model evidence text")
    diff_ops: List[Dict] = Field(default_factory=list, description="Character/token diff operations")
    confidence: float = Field(default=0.0, description="Finding confidence")
    status: str = Field(default="suspected_error", description="Finding status")
    evidence_sources: List[str] = Field(default_factory=list, description="Evidence sources")
    bbox_refs: List[Dict] = Field(default_factory=list, description="Bounding-box references")
    crop_refs: List[str] = Field(default_factory=list, description="Crop/image references")
    wps_anchor: Dict = Field(default_factory=dict, description="WPS DOCX anchor metadata")
    reason: str = Field(default="", description="Decision reason")
    requires_human_review: bool = Field(default=True, description="Whether a human should review")


class PdfWordAuditResult(BaseModel):
    audit_id: str = Field(..., description="Audit task ID")
    filename: str = Field(..., description="Original PDF filename")
    template_filename: str = Field(..., description="WPS DOCX template filename")
    status: str = Field(default="completed", description="Task status")
    metadata: Dict = Field(default_factory=dict, description="Audit metadata")
    product_report: Dict = Field(default_factory=dict, description="Stable product-facing report summary")
    page_risk_summary: List[Dict] = Field(default_factory=list, description="Per-page risk summary")
    table_summary: Dict = Field(default_factory=dict, description="Table review summary")
    coverage_summary: Dict = Field(default_factory=dict, description="Coverage review summary")
    artifact_manifest: Dict = Field(default_factory=dict, description="Artifact manifest for reviewed DOCX/report/evidence")
    review_task_summary: Dict = Field(default_factory=dict, description="Summary of remaining review tasks")
    human_review_queue: List[Dict] = Field(default_factory=list, description="Human review queue items")
    findings: List[PdfWordAuditFinding] = Field(default_factory=list, description="Full OCR audit findings")
    corrections: List[PdfWordAuditCorrection] = Field(default_factory=list, description="Correction and review candidates")
    download_url: Optional[str] = Field(default=None, description="Audited DOCX download URL")
    report_url: Optional[str] = Field(default=None, description="Full JSON report URL")
    evidence_url: Optional[str] = Field(default=None, description="Evidence ZIP URL")
    output_filename: Optional[str] = Field(default=None, description="Audited DOCX filename")


class BatchFileItem(BaseModel):
    item_id: str = Field(..., description="Batch item ID")
    filename: str = Field(..., description="Original filename")
    relative_path: str = Field(..., description="Relative path within the selected folder")
    status: str = Field(..., description="Processing status")
    progress: int = Field(default=0, description="Item progress from 0 to 100")
    message: Optional[str] = Field(default=None, description="Status message")
    error_message: Optional[str] = Field(default=None, description="Terminal error message")
    entities_count: int = Field(default=0, description="Detected entity count")
    output_filename: Optional[str] = Field(default=None, description="Generated output filename")
    mapping_output_filename: Optional[str] = Field(default=None, description="Generated mapping filename")
    output_file_type: Optional[str] = Field(default=None, description="Generated output file type")
    mapping_output_file_type: Optional[str] = Field(default=None, description="Generated mapping file type")
    preserves_format: bool = Field(default=False, description="Whether the original format was preserved")
    warning: Optional[str] = Field(default=None, description="Export warning")
    download_url: Optional[str] = Field(default=None, description="Single-file download URL")
    mapping_download_url: Optional[str] = Field(default=None, description="Mapping document download URL")
    metadata: Dict = Field(default_factory=dict, description="Per-file processing metadata")


class BatchTaskStatus(BaseModel):
    batch_id: str = Field(..., description="Batch task ID")
    folder_name: str = Field(..., description="Selected folder name")
    output_folder_name: Optional[str] = Field(default=None, description="Generated output folder name")
    status: str = Field(..., description="Batch status")
    progress: int = Field(default=0, description="Batch progress from 0 to 100")
    message: Optional[str] = Field(default=None, description="Status message")
    error_message: Optional[str] = Field(default=None, description="Terminal error message")
    file_count: int = Field(default=0, description="Total file count in the batch")
    completed_count: int = Field(default=0, description="Number of finished files")
    succeeded_count: int = Field(default=0, description="Number of successful files")
    failed_count: int = Field(default=0, description="Number of failed files")
    created_at: datetime = Field(default_factory=datetime.now, description="Creation time")
    items: List[BatchFileItem] = Field(default_factory=list, description="Per-file status entries")


class BatchResult(BaseModel):
    batch_id: str = Field(..., description="Batch task ID")
    folder_name: str = Field(..., description="Selected folder name")
    output_folder_name: Optional[str] = Field(default=None, description="Generated output folder name")
    status: str = Field(..., description="Batch status")
    progress: int = Field(default=100, description="Batch progress from 0 to 100")
    message: Optional[str] = Field(default=None, description="Status message")
    error_message: Optional[str] = Field(default=None, description="Terminal error message")
    file_count: int = Field(default=0, description="Total file count in the batch")
    completed_count: int = Field(default=0, description="Number of finished files")
    succeeded_count: int = Field(default=0, description="Number of successful files")
    failed_count: int = Field(default=0, description="Number of failed files")
    archive_download_url: Optional[str] = Field(default=None, description="ZIP archive download URL")
    archive_filename: Optional[str] = Field(default=None, description="ZIP archive filename")
    items: List[BatchFileItem] = Field(default_factory=list, description="Per-file processing results")


class LLMModelOption(BaseModel):
    name: str = Field(..., description="Model name")
    installed: bool = Field(default=False, description="Whether the model is installed locally")
    is_default: bool = Field(default=False, description="Whether this model is the backend default")
    strategy_key: str = Field(..., description="Strategy key applied to this model")
    strategy_label: str = Field(..., description="Strategy label applied to this model")
    strategy_description: str = Field(..., description="Strategy description applied to this model")
    tier: str = Field(default="standard", description="Model capability tier")
    supports_precision_review: bool = Field(default=False, description="Whether the model is suitable for review generation")
    supports_vision: bool = Field(default=False, description="Whether the model can be used for OCR/image-backed workflows")
    recommended_for: List[str] = Field(default_factory=list, description="Recommended usage scenarios")
    role: Optional[str] = Field(default=None, description="Model role in a recognition profile")
    memory_tier: Optional[str] = Field(default=None, description="Estimated memory tier")
    local_path: Optional[str] = Field(default=None, description="Resolved local model path")


class LLMModelListResponse(BaseModel):
    backend: str = Field(..., description="Configured LLM backend")
    default_model: Optional[str] = Field(default=None, description="Backend default LLM model")
    service_available: bool = Field(default=False, description="Whether the backend service is reachable")
    models: List[LLMModelOption] = Field(default_factory=list, description="Selectable models")
    profile: Optional[str] = Field(default=None, description="Recognition profile")
    primary_models: List[str] = Field(default_factory=list, description="Primary local extraction models")
    review_models: List[str] = Field(default_factory=list, description="Fragment review models")


class RuntimeStatusResponse(BaseModel):
    backend: str = Field(..., description="Configured LLM backend")
    platform: str = Field(..., description="Runtime platform label")
    ready: bool = Field(default=False, description="Whether the runtime is ready for full-quality processing")
    ollama_install_detected: bool = Field(default=False, description="Whether an Ollama installation was detected")
    ollama_path: Optional[str] = Field(default=None, description="Detected Ollama installation path")
    service_available: bool = Field(default=False, description="Whether the Ollama service is reachable")
    required_model: Optional[str] = Field(default=None, description="Required local model name")
    required_model_installed: bool = Field(default=False, description="Whether the required model is installed")
    available_processing_models: List[str] = Field(
        default_factory=list,
        description="Installed models that can be used for document processing",
    )
    preferred_processing_model: Optional[str] = Field(
        default=None,
        description="Recommended installed model for the current processing flow",
    )
    default_model: Optional[str] = Field(default=None, description="Default runtime model")
    installer_hint: str = Field(default="", description="Suggested startup hint for the current platform")
    download_hint: str = Field(default="", description="Suggested model download command or script")
    recommended_action: str = Field(default="", description="Single-sentence recommended next step")
    desensitize_mode: Optional[str] = Field(default=None, description="Configured desensitization mode")
    primary_models_ready: Optional[bool] = Field(default=None, description="Whether primary low-memory models are ready")
    review_model_installed: Optional[bool] = Field(default=None, description="Whether fragment review model is installed")
    review_model_loaded: Optional[bool] = Field(default=None, description="Whether fragment review model is currently loaded")
    estimated_memory_tier: Optional[str] = Field(default=None, description="Estimated memory tier")
    analysis_worker_process: Optional[bool] = Field(default=None, description="Whether high-memory analysis runs in one-shot workers")
    analysis_stage_isolation: Optional[bool] = Field(default=None, description="Whether high-memory stages run in sequential isolated workers")
    analysis_worker_timeout: Optional[int] = Field(default=None, description="Analysis worker timeout in seconds")


class PageSessionRequest(BaseModel):
    page_session_id: str = Field(..., description="Frontend page session ID")


class PageSessionResponse(BaseModel):
    ok: bool = Field(default=True, description="Whether the request was accepted")

